model/test_results_iterator.go

```diff
@@ -49,47 +57,103 @@ func (i *testResultsIterator) Next(ctx context.Context) bool {
 		}
 	}()
 
-	if i.iter == nil {
+	if len(i.bucketItems) == 0 {
 		iter, err := i.bucket.List(ctx, "")
 		if err != nil {
 			i.catcher.Wrap(err, "listing bucket contents")
 			return false
 		}
-		i.iter = iter
+
+		for iter.Next(ctx) {
+			i.bucketItems = append(i.bucketItems, iter.Item())
+		}
+		if err = iter.Err(); err != nil {
+			i.catcher.Wrapf(err, "iterating bucket contents")
+			return false
+		}
 	}
-	if !i.iter.Next(ctx) {
-		i.catcher.Wrapf(i.iter.Err(), "iterating bucket contents")
-		i.exhausted = true
-		return false
+
+	if len(i.items) == 0 {
+		if i.currentIdx == len(i.bucketItems) {
+			i.exhausted = true
+			return false
+		}
+
+		if err := i.getNextBatch(ctx); err != nil {
+			i.catcher.Add(err)
+			return false
+		}
 	}
 
-	r, err := i.iter.Item().Get(ctx)
-	if err != nil {
-		i.catcher.Wrapf(err, "item '%s'", i.iter.Item().Name())
+	select {
+	case <-ctx.Done():
+		i.catcher.Add(ctx.Err())
 		return false
+	case i.currentItem = <-i.items:
+		return true
 	}
-	defer func() {
-		grip.Error(message.WrapError(r.Close(), message.Fields{
-			"message": "could not close bucket reader",
-			"bucket":  i.bucket,
-			"path":    i.iter.Item().Name(),
-		}))
-	}()
+}
 
-	data, err := ioutil.ReadAll(r)
-	if err != nil {
-		i.catcher.Wrapf(err, "reading test result '%s'", i.iter.Item().Name())
-		return false
+func (i *testResultsIterator) getNextBatch(ctx context.Context) error {
+	idxs := make(chan int, testResultsIteratorBatchSize)
+	for j := 0; j < testResultsIteratorBatchSize; j++ {
+		if i.currentIdx == len(i.bucketItems) {
+			break
+		}
+		idxs <- i.currentIdx
+		i.currentIdx++
 	}
-	var result TestResult
-	if err := bson.Unmarshal(data, &result); err != nil {
-		i.catcher.Wrapf(err, "unmarshalling test result '%s'", i.iter.Item().Name())
-		return false
+	close(idxs)
+	var wg sync.WaitGroup
+	catcher := grip.NewBasicCatcher()
+
+	for j := 0; j < runtime.NumCPU(); j++ {
+		wg.Add(1)
+		go func() {
+			defer func() {
+				var err error
+				catcher.Add(recovery.HandlePanicWithError(recover(), err))
```

Comment:
Can we add a message as the 3rd argument to `recovery.HandlePanicWithError` so we have context for the source of the error (e.g. "iterating test results in worker")?